### PR TITLE
schemachanger: implement composite types

### PR DESCRIPTION
**Pull request by jordanlewis**
Based on #90491 

This commit adds composite type support to the new schema changer.

Epic: CRDB-22358

Release note: None

### CODE REVIEW

**Comment by jordanlewis** [comment]
Hey @postamar, here is this implementation of composite types in the new schema changer.

File: pkg/sql/schemachanger/scplan/internal/opgen/opgen_composite_type_attribute.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+				emit(func(this *scpb.CompositeTypeAttribute) *scop.UpdateTypeBackReferencesInTypes {
+					if len(this.ClosedTypeIDs) == 0 {
+						return nil
+					}
+					var ids catalog.DescriptorIDSet
+					for _, id := range this.ClosedTypeIDs {
+						ids.Add(id)
+					}
+					return &scop.UpdateTypeBackReferencesInTypes{
+						TypeIDs:              ids.Ordered(),
```
Comment (postamar) [inline pkg/sql/schemachanger/scplan/internal/opgen/opgen_composite_type_attribute.go:33]: You should just plop `this.ClosedTypeIDs` in there after checking that it's not empty. It should be considered as part of `scdecomp`'s contract that the `TypeT` messages are well-formed. Same comment applies below.

**Review by postamar — CHANGES_REQUESTED** [review]
I just have one major change request related to the element model, but performing it should be straightforward and I look forward to approving this PR soon. This otherwise looks quite sane and if CI's happy then I'm happy.

File: pkg/sql/schemachanger/scpb/elements.proto [hunk 2]
```diff
@@ -506,6 +516,12 @@ message EnumTypeValue {
   string logical_representation = 3;
 }
 
+message CompositeTypeAttribute {
+  uint32 type_id = 1 [(gogoproto.customname) = "TypeID", (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/sql/sem/catid.DescID"];
+  string label = 2;
```
[the remark this answers is not in the document]
Comment (jordanlewis) [reply pkg/sql/schemachanger/scpb/elements.proto]: I see, yes. Fixed, I think - PTAL.

File: pkg/sql/schemachanger/scplan/internal/opgen/opgen_composite_type_attribute.go (hunk 1 above)
Comment (jordanlewis) [reply pkg/sql/schemachanger/scplan/internal/opgen/opgen_composite_type_attribute.go]: Nice, got it, Done.

**Comment by jordanlewis** [comment]
@postamar could you take another look? Rebased, and I think all of your comments are resolved.

**Review by postamar — APPROVED** [review]
LGTM. Sorry this took so long! I should have reviewed this last week.

**Comment by jordanlewis** [comment]
TFTR!

bors r+

**Pull request merged**